File: desktop/quiz.py
"""Quiz questions and the game state machine.

Questions are multiple choice with up to four options. The player answers
by holding up the matching number of fingers (1-4) in front of the webcam.
An answer is only accepted once the same finger count has been stable for
STABLE_SECONDS (time-based, so it feels identical on fast/slow machines
and is forgiving for slow or trembling hands).

NOTE: the web version (quiz.mjs) uses the same questions in Hindi
(Devanagari); this desktop copy uses English translations because OpenCV's
built-in font cannot render Devanagari.

This module is pure Python (no cv2 / mediapipe), so the whole game flow can
be unit-tested without a camera.
"""

import random


class Question:
    def __init__(self, prompt, options, answer):
        self.prompt = prompt
        self.options = options
        self.answer = answer  # 1-based index into options

    @property
    def correct_text(self):
        return self.options[self.answer - 1]


# Answer with: 1 finger, 2 fingers, 3 fingers or 4 fingers.
# English translations of the Hindi question bank in quiz.mjs (web).
QUESTION_BANK = [
    Question("What is the capital of India?",
              ["Mumbai", "New Delhi", "Kolkata", "Chennai"], answer=2),
    Question("Which is the national bird of India?",
              ["Parrot", "Pigeon", "Swan", "Peacock"], answer=4),
    Question("In which city is the Taj Mahal located?",
              ["Jaipur", "Agra", "Delhi", "Lucknow"], answer=2),
    Question("Who had the Taj Mahal built?",
              ["Akbar", "Shah Jahan", "Aurangzeb", "Humayun"], answer=2),
    Question("What is the national anthem of India?",
              ["Vande Mataram", "Sare Jahan Se Achha", "Jana Gana Mana", "Jai Jawan Jai Kisan"], answer=3),
    Question("Which colour is at the top of the Indian flag?",
              ["Green", "White", "Saffron", "Blue"], answer=3),
    Question("Who was the first Prime Minister of India?",
              ["Mahatma Gandhi", "Sardar Patel", "Lal Bahadur Shastri", "Jawaharlal Nehru"], answer=4),
    Question("Which is the longest river of India?",
              ["Ganga", "Yamuna", "Godavari", "Kaveri"], answer=1),
    Question("Which is the national animal of India?",
              ["Lion", "Elephant", "Tiger", "Cow"], answer=3),
    Question("Diwali is the festival of what?",
              ["Colours", "Rakhi", "Kites", "Lamps and light"], answer=4),
    Question("How many balls are bowled in one over of cricket?",
              ["4", "5", "6", "8"], answer=3),
    Question("Sachin Tendulkar is a legend of which sport?",
              ["Hockey", "Football", "Tennis", "Cricket"], answer=4),
    Question("In which year did Chandrayaan-3 land on the Moon?",
              ["2019", "2021", "2022", "2023"], answer=4),
    Question("UPI is the digital payment system of which country?",
              ["Nepal", "India", "Sri Lanka", "Japan"], answer=2),
    Question("Holi is famous as the festival of what?",
              ["Colours", "Lamps", "Sweets", "Fasting"], answer=1),
    Question("Which is the highest mountain peak in the world?",
              ["K2", "Kangchenjunga", "Mount Everest", "Makalu"], answer=3),
    Question("In how many hours does the Earth complete one rotation?",
              ["12", "24", "36", "48"], answer=2),
    Question("Which is the largest ocean in the world?",
              ["Atlantic", "Indian", "Arctic", "Pacific"], answer=4),
    Question("In which direction does the sun rise?",
              ["East", "West", "North", "South"], answer=1),
    Question("Who discovered the law of gravity?",
              ["Albert Einstein", "Galileo", "Isaac Newton", "Charles Darwin"], answer=3),
]


# --- Game phases ------------------------------------------------------------
INTRO = "intro"        # waiting for the player to show a hand
QUESTION = "question"  # waiting for a stable finger count
FEEDBACK = "feedback"  # showing right/wrong for a moment
DONE = "done"         # final score screen

STABLE_SECONDS = 0.8   # seconds a finger count must be held (time-based)
FEEDBACK_SECONDS = 2.0


class QuizGame:
    """Drives the quiz; feed it detections, read out what to draw."""

    def __init__(self, questions=None, seed=None):
        self.rng = random.Random(seed)
        self.all_questions = list(questions if questions is not None else QUESTION_BANK)
        self.elapsed = 0.0
        self.restart()

    # -- setup / lifecycle ---------------------------------------------------
    def restart(self):
        self.questions = self.all_questions[:]
        self.rng.shuffle(self.questions)
        self.index = 0
        self.score = 0
        self.phase = INTRO
        self.elapsed = 0.0
        self._last_count = 0
        self._hold_time = 0.0

    @property
    def current_question(self):
        if 0 <= self.index < len(self.questions):
            return self.questions[self.index]
        return None

    @property
    def total(self):
        return len(self.questions)

    # -- main update ----------------------------------------------------------
    def update(self, finger_counts, dt):
        """Advance the game one frame.

        finger_counts: list of per-hand finger counts detected this frame
                       (empty list when no hand is visible).
        dt: seconds since the previous frame.
        """
        self.elapsed += dt
        best = max(finger_counts) if finger_counts else 0
        if best == self._last_count:
            self._hold_time += dt
        else:
            self._last_count = best
            self._hold_time = 0.0
        stable = self._hold_time >= STABLE_SECONDS

        if self.phase == INTRO:
            if best >= 1 and stable:
                self._start_question()

        elif self.phase == QUESTION:
            if 1 <= best <= 4 and stable:
                self._submit_answer(best)

        elif self.phase == FEEDBACK:
            if self.elapsed >= FEEDBACK_SECONDS:
                self.index += 1
                if self.index >= len(self.questions):
                    self.phase = DONE
                    self.elapsed = 0.0
                else:
                    self._start_question()

        elif self.phase == DONE:
            if best >= 1 and stable:
                self.restart()

    # -- helpers ---------------------------------------------------------------
    def _start_question(self):
        self.phase = QUESTION
        self.elapsed = 0.0
        self._last_count = 0
        self._hold_time = 0.0

    def _submit_answer(self, answer):
        q = self.current_question
        self.last_answer = answer
        self.last_correct = (answer == q.answer)
        if self.last_correct:
            self.score += 1
        self.phase = FEEDBACK
        self.elapsed = 0.0

    # -- things the UI likes to know -------------------------------------------
    @property
    def held_count(self):
        """The finger count currently being held (for the progress bar)."""
        return self._last_count if self._last_count >= 1 else 0

    @property
    def lock_progress(self):
        """0.0 - 1.0, how close the held gesture is to being accepted."""
        if self._last_count < 1:
            return 0.0
        return min(1.0, self._hold_time / STABLE_SECONDS)

File: quiz.mjs
/* Hand Quiz — question bank + game state machine.
   Questions are in Hindi (Devanagari): 15 about India + 5 about the world,
   mostly easy — designed to be read by everyone including elderly players.

   The lock-in is TIME-based (~0.8 s) instead of frame-based, so it feels
   the same on fast and slow devices, and is forgiving for slow or
   trembling hands.

   No DOM / CDN dependencies, so it can be unit-tested with plain Node:
   node quiz.test.mjs */

export const STABLE_SECONDS = 0.8; // seconds a finger count must be held
export const FEEDBACK_SECONDS = 2.0; // a little longer, for comfortable reading

export const QUESTIONS = [
  { prompt: "भारत की राजधानी क्या है?",
    options: ["मुंबई", "नई दिल्ली", "कोलकाता", "चेन्नई"], answer: 2 },
  { prompt: "भारत का राष्ट्रीय पक्षी कौन सा है?",
    options: ["तोता", "कबूतर", "हंस", "मोर"], answer: 4 },
  { prompt: "ताजमहल किस शहर में स्थित है?",
    options: ["जयपुर", "आगरा", "दिल्ली", "लखनऊ"], answer: 2 },
  { prompt: "ताजमहल का निर्माण किसने करवाया था?",
    options: ["अकबर", "शाहजहाँ", "औरंगज़ेब", "हुमायूँ"], answer: 2 },
  { prompt: "भारत का राष्ट्रगान क्या है?",
    options: ["वंदे मातरम्", "सारे जहाँ से अच्छा", "जन गण मन", "जय जवान जय किसान"], answer: 3 },
  { prompt: "भारतीय ध्वज में सबसे ऊपर कौन सा रंग होता है?",
    options: ["हरा", "सफेद", "केसरिया", "नीला"], answer: 3 },
  { prompt: "भारत के पहले प्रधानमंत्री कौन थे?",
    options: ["महात्मा गांधी", "सरदार पटेल", "लाल बहादुर शास्त्री", "जवाहरलाल नेहरू"], answer: 4 },
  { prompt: "भारत की सबसे लंबी नदी कौन सी है?",
    options: ["गंगा", "यमुना", "गोदावरी", "कावेरी"], answer: 1 },
  { prompt: "भारत का राष्ट्रीय पशु कौन सा है?",
    options: ["शेर", "हाथी", "बाघ", "गाय"], answer: 3 },
  { prompt: "दिवाली किस चीज़ का त्योहार है?",
    options: ["रंगों का", "राखी का", "पतंगों का", "दीपों और रोशनी का"], answer: 4 },
  { prompt: "क्रिकेट के एक ओवर में कितनी गेंदें फेंकी जाती हैं?",
    options: ["4", "5", "6", "8"], answer: 3 },
  { prompt: "सचिन तेंदुलकर किस खेल के महान खिलाड़ी हैं?",
    options: ["हॉकी", "फुटबॉल", "टेनिस", "क्रिकेट"], answer: 4 },
  { prompt: "चंद्रयान-3 चाँद पर किस वर्ष सफलतापूर्वक उतरा?",
    options: ["2019", "2021", "2022", "2023"], answer: 4 },
  { prompt: "UPI डिजिटल भुगतान प्रणाली किस देश की है?",
    options: ["नेपाल", "भारत", "श्रीलंका", "जापान"], answer: 2 },
  { prompt: "होली का त्योहार किस चीज़ के लिए प्रसिद्ध है?",
    options: ["रंगों के लिए", "दीयों के लिए", "मिठाई के लिए", "उपवास के लिए"], answer: 1 },
  { prompt: "विश्व की सबसे ऊँची पर्वत चोटी कौन सी है?",
    options: ["के-टू", "कंचनजंगा", "माउंट एवरेस्ट", "मकालू"], answer: 3 },
  { prompt: "पृथ्वी अपनी धुरी पर एक चक्कर कितने घंटों में पूरा करती है?",
    options: ["12", "24", "36", "48"], answer: 2 },
  { prompt: "विश्व का सबसे बड़ा महासागर कौन सा है?",
    options: ["अटलांटिक", "हिंद", "आर्कटिक", "प्रशांत"], answer: 4 },
  { prompt: "सूर्य किस दिशा में उगता है?",
    options: ["पूर्व", "पश्चिम", "उत्तर", "दक्षिण"], answer: 1 },
  { prompt: "गुरुत्वाकर्षण का नियम किसने खोजा था?",
    options: ["अल्बर्ट आइंस्टीन", "गैलीलियो", "आइज़क न्यूटन", "चार्ल्स डार्विन"], answer: 3 },
];

export class QuizGame {
  constructor(questions = QUESTIONS) {
    this.allQuestions = questions;
    this.restart();
  }

  restart() {
    this.questions = [...this.allQuestions].sort(() => Math.random() - 0.5);
    this.index = 0;
    this.score = 0;
    this.phase = "intro"; // intro -> question -> feedback -> done
    this.elapsed = 0;
    this.lastAnswer = null;
    this.lastCorrect = null;
    this._lastCount = 0;
    this._holdTime = 0;
  }

  get currentQuestion() {
    return this.questions[this.index];
  }

  get total() {
    return this.questions.length;
  }

  update(counts, dt) {
    this.elapsed += dt;
    const best = counts.length ? Math.max(...counts) : 0;
    if (best === this._lastCount) {
      this._holdTime += dt;
    } else {
      this._lastCount = best;
      this._holdTime = 0;
    }
    const stable = this._holdTime >= STABLE_SECONDS;

    if (this.phase === "intro") {
      if (best >= 1 && stable) this._startQuestion();
    } else if (this.phase === "question") {
      if (best >= 1 && best <= 4 && stable) this._submit(best);
    } else if (this.phase === "feedback") {
      if (this.elapsed >= FEEDBACK_SECONDS) {
        this.index += 1;
        this.elapsed = 0;
        if (this.index >= this.questions.length) {
          this.phase = "done";
        } else {
          this._startQuestion();
        }
      }
    } else if (this.phase === "done") {
      if (best >= 1 && stable) this.restart();
    }
  }

  _startQuestion() {
    this.phase = "question";
    this.elapsed = 0;
    this._lastCount = 0;
    this._holdTime = 0;
  }

  _submit(answer) {
    this.lastAnswer = answer;
    this.lastCorrect = answer === this.currentQuestion.answer;
    if (this.lastCorrect) this.score += 1;
    this.phase = "feedback";
    this.elapsed = 0;
  }

  heldCount() {
    return this._lastCount >= 1 ? this._lastCount : 0;
  }

  lockProgress() {
    if (this._lastCount < 1) return 0;
    return Math.min(1, this._holdTime / STABLE_SECONDS);
  }
}
